pdns/protozero.hh

```diff
@@ -0,0 +1,254 @@
+/*
+ * This file is part of PowerDNS or dnsdist.
+ * Copyright -- PowerDNS.COM B.V. and its contributors
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of version 2 of the GNU General Public License as
+ * published by the Free Software Foundation.
+ *
+ * In addition, for the avoidance of any doubt, permission is granted to
+ * link this program with OpenSSL and to (re)distribute the binaries
+ * produced as the result of such linking.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+#pragma once
+
+#include <protozero/pbf_writer.hpp>
+#include <string>
+
+#include "config.h"
+#include "iputils.hh"
+#include "filterpo.hh"
+#include "gettime.hh"
+#include "uuid-utils.hh"
+
+namespace pdns {
+  namespace ProtoZero {
+    class Message {
+    public:
+      // Start a new messagebug, constaining separate data for the response part
+      Message(std::string::size_type sz1, std::string::size_type sz2) : d_message{d_msgbuf}, d_response{d_rspbuf}
+      {
+        // This is extra space in addition to what's already there
+        // Different from what string.reserve() does
+        d_message.reserve(sz1);
```

Comment:
Be careful when reserving a `std::string`, it can actually trigger a reallocation of a smaller buffer (before C++20). I would check the `capacity()` before calling `reserve()`.

Reply:
This is calling reserve on `protozero::pbf_writer` which has different semantics compared to `std::string`.

Reply:
Right, but the issue is the same because that does call `std::string::reserve()` in the end, adding the new size to the existing size, not to the existing capacity. That doesn't matter now, but if we end up reusing an existing string to avoid allocations we sometimes will cause a reallocation by calling reserve() with a smaller size than the current capacity of the string.